provider/ec2/environ.go

```diff
@@ -2530,7 +2537,24 @@ func (e *environ) SetCloudSpec(spec environscloudspec.CloudSpec) error {
 		return errors.Trace(err)
 	}
 
-	e.ec2Client = EC2Session(e.cloud.Region, e.ec2.AccessKey, e.ec2.SecretKey)
+	// Allow the passing of a client func through the context. This allows
+	// passing the client from outside of the environ, one that allows for
+	// custom http.Clients.
+	if value := ctx.Value(AWSClientContextKey); value == nil {
```

Comment:
This seems like a horrible hack to me :-(

I really wish we could just take the time and move all tests into the ec2 package so we could simply call the constructor as a proper unit-test would do...

Reply:
I see this as one step forward, but necessarily the final step. I'm not that enamored in how we create an environ, to begin with. I think if we can keep iterating on this a little more, I think we will get to a place of just passing in a value via the constructor.

Reply:
https://github.com/juju/juju/pull/13077/commits/cb4f27d285fe892b253b254ce3271c25d8d443aa